File: zbudget/lib/models/challenge.dart
import 'package:flutter/material.dart';

enum ChallengeDifficulty { easy, medium, hard }

enum ChallengeStatus { available, active, completed }

enum ChallengeCategory {
  drinks,
  food,
  transport,
  shopping,
  entertainment,
  special,
  digital,
}

class ChallengeProgress {
  final String id;
  final String challengeId;
  final String userId;
  final DateTime date;
  final double amountSaved;
  final String note;
  final List<String> proofImages;
  final bool isVerified;

  ChallengeProgress({
    required this.id,
    required this.challengeId,
    required this.userId,
    required this.date,
    required this.amountSaved,
    this.note = '',
    this.proofImages = const [],
    this.isVerified = false,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'challengeId': challengeId,
      'userId': userId,
      'date': date.toIso8601String(),
      'amountSaved': amountSaved,
      'note': note,
      'proofImages': proofImages,
      'isVerified': isVerified,
    };
  }

  factory ChallengeProgress.fromJson(Map<String, dynamic> json) {
    return ChallengeProgress(
      id: json['id'],
      challengeId: json['challengeId'],
      userId: json['userId'],
      date: DateTime.parse(json['date']),
      amountSaved: json['amountSaved'].toDouble(),
      note: json['note'] ?? '',
      proofImages: List<String>.from(json['proofImages'] ?? []),
      isVerified: json['isVerified'] ?? false,
    );
  }
}

class ChallengeMilestone {
  final int day;
  final double targetAmount;
  final String reward;
  final bool isCompleted;
  final String description;

  ChallengeMilestone({
    required this.day,
    required this.targetAmount,
    required this.reward,
    this.isCompleted = false,
    required this.description,
  });

  Map<String, dynamic> toJson() {
    return {
      'day': day,
      'targetAmount': targetAmount,
      'reward': reward,
      'isCompleted': isCompleted,
      'description': description,
    };
  }

  factory ChallengeMilestone.fromJson(Map<String, dynamic> json) {
    return ChallengeMilestone(
      day: json['day'],
      targetAmount: json['targetAmount'].toDouble(),
      reward: json['reward'],
      isCompleted: json['isCompleted'] ?? false,
      description: json['description'],
    );
  }
}

class Challenge {
  final String id;
  final String emoji;
  final String title;
  final String description;
  final int participants;
  final String duration;
  final int amount;
  final ChallengeDifficulty difficulty;
  final int points;
  final ChallengeStatus status;
  final ChallengeCategory category;
  final int? progress;
  final DateTime? startDate;
  final List<ChallengeMilestone> milestones;
  final double? currentSaved;
  final int? streakDays;
  final DateTime? endDate;
  final int? currentSavings;
  final int? streak;
  final String? badgeIcon;
  final DateTime? completedAt;

  Challenge({
    required this.id,
    required this.emoji,
    required this.title,
    required this.description,
    required this.participants,
    required this.duration,
    required this.amount,
    required this.difficulty,
    required this.points,
    required this.status,
    required this.category,
    this.progress,
    this.startDate,
    this.endDate,
    this.currentSavings,
    this.streak,
    this.badgeIcon,
    this.completedAt,
    this.milestones = const [],
    this.currentSaved,
    this.streakDays,
  });

  Challenge copyWith({
    String? id,
    String? emoji,
    String? title,
    String? description,
    int? participants,
    String? duration,
    int? amount,
    ChallengeDifficulty? difficulty,
    int? points,
    ChallengeStatus? status,
    ChallengeCategory? category,
    int? progress,
    DateTime? startDate,
    DateTime? endDate,
    int? currentSavings,
    int? streak,
    String? badgeIcon,
    DateTime? completedAt,
  }) {
    return Challenge(
      id: id ?? this.id,
      emoji: emoji ?? this.emoji,
      title: title ?? this.title,
      description: description ?? this.description,
      participants: participants ?? this.participants,
      duration: duration ?? this.duration,
      amount: amount ?? this.amount,
      difficulty: difficulty ?? this.difficulty,
      points: points ?? this.points,
      status: status ?? this.status,
      category: category ?? this.category,
      progress: progress ?? this.progress,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      currentSavings: currentSavings ?? this.currentSavings,
      streak: streak ?? this.streak,
      badgeIcon: badgeIcon ?? this.badgeIcon,
      completedAt: completedAt ?? this.completedAt,
    );
  }
}

class UserStats {
  final int streak;
  final int level;
  final int points;
  final int totalChallengesCompleted;
  final int totalSavings;
  final int rank;
  final int nextLevelPoints;

  UserStats({
    required this.streak,
    required this.level,
    required this.points,
    required this.totalChallengesCompleted,
    required this.totalSavings,
    required this.rank,
    required this.nextLevelPoints,
  });
}

class ChallengeCategoryModel {
  final String id;
  final String name;
  final String icon;
  final Color color;

  ChallengeCategoryModel({
    required this.id,
    required this.name,
    required this.icon,
    required this.color,
  });
}

File: Backend/models/Challenge.js
import mongoose from "mongoose";
import validator from "validator";

// Sub-schemas
const DurationSchema = new mongoose.Schema(
  {
    days: {
      type: Number,
      min: [1, "Thời gian thách thức phải ít nhất 1 ngày"],
      max: [365, "Thời gian thách thức không được quá 365 ngày"],
      required: [true, "Số ngày là bắt buộc"],
    },
    displayText: {
      type: String,
      maxlength: [50, "Hiển thị thời gian không được vượt quá 50 ký tự"],
    },
    timeLimit: {
      type: Date, // null = no time limit to start
    },
  },
  { _id: false }
);

const TargetsSchema = new mongoose.Schema(
  {
    estimatedSaving: {
      type: mongoose.Schema.Types.Decimal128,
      required: [true, "Mục tiêu tiết kiệm dự kiến là bắt buộc"],
      validate: {
        validator: function (v) {
          return parseFloat(v.toString()) >= 0;
        },
        message: "Mục tiêu tiết kiệm phải >= 0",
      },
    },
    dailySavingTarget: {
      type: mongoose.Schema.Types.Decimal128,
      default: 0,
    },
    maxBudget: {
      type: mongoose.Schema.Types.Decimal128,
      default: 0,
    },
    comparisonPeriod: {
      type: String,
      enum: ["previous_week", "previous_month", "average", "custom"],
      default: "previous_week",
    },
  },
  { _id: false }
);

const MilestoneSchema = new mongoose.Schema(
  {
    day: {
      type: Number,
      min: [1, "Ngày milestone phải >= 1"],
      required: [true, "Ngày milestone là bắt buộc"],
    },
    title: {
      type: String,
      required: [true, "Tiêu đề milestone là bắt buộc"],
      maxlength: [100, "Tiêu đề milestone không được vượt quá 100 ký tự"],
    },
    description: {
      type: String,
      maxlength: [200, "Mô tả milestone không được vượt quá 200 ký tự"],
    },
    points: {
      type: Number,
      min: [0, "Điểm milestone phải >= 0"],
      required: [true, "Điểm milestone là bắt buộc"],
    },
    badge: {
      type: String,
      maxlength: [10, "Badge không được vượt quá 10 ký tự"],
    },
    reward: {
      type: String,
      maxlength: [200, "Phần thưởng không được vượt quá 200 ký tự"],
    },
  },
  { _id: false }
);

const StatsSchema = new mongoose.Schema(
  {
    totalParticipants: {
      type: Number,
      min: 0,
      default: 0,
    },
    activeParticipants: {
      type: Number,
      min: 0,
      default: 0,
    },
    completionRate: {
      type: Number,
      min: 0,
      max: 100,
      default: 0,
    },
    averageSaving: {
      type: mongoose.Schema.Types.Decimal128,
      default: 0,
    },
    topSaver: {
      type: mongoose.Schema.Types.Decimal128,
      default: 0,
    },
  },
  { _id: false }
);

const RulesSchema = new mongoose.Schema(
  {
    allowedExpenses: [
      {
        type: String,
        maxlength: [50, "Mục chi tiêu cho phép không được vượt quá 50 ký tự"],
      },
    ],
    forbiddenCategories: [
      {
        type: String,
        enum: [
          "food",
          "transport",
          "shopping",
          "entertainment",
          "healthcare",
          "education",
          "utilities",
          "other",
          "coffee-shops",
          "cafe",
          "drinks",
        ],
      },
    ],
    substitutionSuggestions: [
      {
        type: String,
        maxlength: [200, "Gợi ý thay thế không được vượt quá 200 ký tự"],
      },
    ],
    trackingMethod: {
      type: String,
      enum: [
        "expense_category",
        "total_amount",
        "manual_logging",
        "receipt_scan",
      ],
      default: "expense_category",
    },
  },
  { _id: false }
);

// Main Challenge Schema
const ChallengeSchema = new mongoose.Schema(
  {
    // Challenge Identity
    challengeId: {
      type: String,
      required: [true, "Challenge ID là bắt buộc"],
      unique: true,
      lowercase: true,
      trim: true,
      maxlength: [100, "Challenge ID không được vượt quá 100 ký tự"],
      validate: {
        validator: function (v) {
          return /^[a-z0-9-_]+$/.test(v);
        },
        message:
          "Challenge ID chỉ được chứa chữ thường, số, dấu gạch ngang và gạch dưới",
      },
      index: true,
    },
    title: {
      type: String,
      required: [true, "Tiêu đề thách thức là bắt buộc"],
      trim: true,
      maxlength: [200, "Tiêu đề không được vượt quá 200 ký tự"],
    },
    description: {
      type: String,
      required: [true, "Mô tả thách thức là bắt buộc"],
      trim: true,
      maxlength: [1000, "Mô tả không được vượt quá 1000 ký tự"],
    },

    // Visual & Branding
    emoji: {
      type: String,
      maxlength: [10, "Emoji không được vượt quá 10 ký tự"],
    },
    color: {
      type: String,
      validate: {
        validator: function (v) {
          return !v || /^#[0-9A-F]{6}$/i.test(v);
        },
        message: "Màu sắc phải là hex color hợp lệ",
      },
    },
    bannerImage: {
      type: String,
      validate: {
        validator: function (v) {
          return !v || validator.isURL(v);
        },
        message: "Banner image phải là URL hợp lệ",
      },
    },

    // Challenge Configuration
    type: {
      type: String,
      enum: ["reduction", "savings", "budget", "custom"],
      default: "reduction",
    },
    category: {
      type: String,
      enum: [
        "drinks",
        "food",
        "transport",
        "shopping",
        "entertainment",
        "lifestyle",
      ],
      required: [true, "Danh mục thách thức là bắt buộc"],
      index: true,
    },
    difficulty: {
      type: String,
      enum: ["easy", "medium", "hard"],
      required: [true, "Độ khó là bắt buộc"],
      index: true,
    },

    // Duration & Timeline
    duration: {
      type: DurationSchema,
      required: [true, "Thời gian thách thức là bắt buộc"],
    },

    // Financial Targets
    targets: {
      type: TargetsSchema,
      required: [true, "Mục tiêu tài chính là bắt buộc"],
    },

    // Milestones & Rewards
    milestones: {
      type: [MilestoneSchema],
      validate: {
        validator: function (v) {
          return v.length > 0;
        },
        message: "Phải có ít nhất 1 milestone",
      },
    },

    // Participation & Stats
    stats: {
      type: StatsSchema,
      default: () => ({}),
    },

    // Rules & Guidelines
    rules: {
      type: RulesSchema,
      default: () => ({}),
    },

    // Challenge Status
    isActive: {
      type: Boolean,
      default: true,
      index: true,
    },
    startDate: {
      type: Date,
      default: Date.now,
    },
    endDate: {
      type: Date,
      required: [true, "Ngày kết thúc thách thức là bắt buộc"],
    },
    featured: {
      type: Boolean,
      default: false,
      index: true,
    },

    // Creator Info
    createdBy: {
      type: String,
      default: "system", // 'system' or userId
    },

    version: {
      type: Number,
      default: 1,
    },
  },
  {
    timestamps: true,
    versionKey: false,
    toJSON: {
      transform: function (doc, ret) {
        // Convert Decimal128 to number for JSON
        if (ret.targets) {
          if (ret.targets.estimatedSaving)
            ret.targets.estimatedSaving = parseFloat(
              ret.targets.estimatedSaving.toString()
            );
          if (ret.targets.dailySavingTarget)
            ret.targets.dailySavingTarget = parseFloat(
              ret.targets.dailySavingTarget.toString()
            );
          if (ret.targets.maxBudget)
            ret.targets.maxBudget = parseFloat(
              ret.targets.maxBudget.toString()
            );
        }
        if (ret.stats) {
          if (ret.stats.averageSaving)
            ret.stats.averageSaving = parseFloat(
              ret.stats.averageSaving.toString()
            );
          if (ret.stats.topSaver)
            ret.stats.topSaver = parseFloat(ret.stats.topSaver.toString());
        }
        return ret;
      },
    },
  }
);

// Indexes
ChallengeSchema.index({ isActive: 1, featured: 1 });
ChallengeSchema.index({ category: 1, difficulty: 1 });
ChallengeSchema.index({ startDate: 1, endDate: 1 });

// Instance Methods
ChallengeSchema.methods.addParticipant = function () {
  this.stats.totalParticipants += 1;
  this.stats.activeParticipants += 1;
};

ChallengeSchema.methods.removeParticipant = function () {
  if (this.stats.activeParticipants > 0) {
    this.stats.activeParticipants -= 1;
  }
};

ChallengeSchema.methods.updateCompletionRate = function () {
  if (this.stats.totalParticipants > 0) {
    const completed =
      this.stats.totalParticipants - this.stats.activeParticipants;
    this.stats.completionRate =
      (completed / this.stats.totalParticipants) * 100;
  }
};

ChallengeSchema.methods.updateAverageSaving = function (
  totalSavings,
  participantCount
) {
  if (participantCount > 0) {
    this.stats.averageSaving = mongoose.Types.Decimal128.fromString(
      (totalSavings / participantCount).toFixed(2)
    );
  }
};

ChallengeSchema.methods.updateTopSaver = function (savingAmount) {
  const currentTop = parseFloat(this.stats.topSaver.toString()) || 0;
  if (savingAmount > currentTop) {
    this.stats.topSaver = mongoose.Types.Decimal128.fromString(
      savingAmount.toFixed(2)
    );
  }
};

ChallengeSchema.methods.getMilestoneByDay = function (day) {
  return this.milestones.find((milestone) => milestone.day === day);
};

ChallengeSchema.methods.isExpired = function () {
  return new Date() > this.endDate;
};

ChallengeSchema.methods.getDaysRemaining = function () {
  const now = new Date();
  const diffTime = this.endDate - now;
  return Math.max(0, Math.ceil(diffTime / (1000 * 60 * 60 * 24)));
};

// Pre-save middleware
ChallengeSchema.pre("save", function (next) {
  // Calculate daily saving target
  if (
    this.isModified("targets.estimatedSaving") ||
    this.isModified("duration.days")
  ) {
    const estimatedSaving = parseFloat(this.targets.estimatedSaving.toString());
    const dailyTarget = estimatedSaving / this.duration.days;
    this.targets.dailySavingTarget = mongoose.Types.Decimal128.fromString(
      dailyTarget.toFixed(2)
    );
  }

  // Update duration display text
  if (this.isModified("duration.days")) {
    const days = this.duration.days;
    if (days === 1) {
      this.duration.displayText = "1 ngày";
    } else if (days === 7) {
      this.duration.displayText = "1 tuần";
    } else if (days === 30) {
      this.duration.displayText = "1 tháng";
    } else {
      this.duration.displayText = `${days} ngày`;
    }
  }

  // Update version
  if (this.isModified() && !this.isNew) {
    this.version += 1;
  }

  next();
});

// Static Methods
ChallengeSchema.statics.findActive = function () {
  const now = new Date();
  return this.find({
    isActive: true,
    startDate: { $lte: now },
    endDate: { $gte: now },
  }).sort({ featured: -1, createdAt: -1 });
};

ChallengeSchema.statics.findFeatured = function (limit = 5) {
  const now = new Date();
  return this.find({
    isActive: true,
    featured: true,
    startDate: { $lte: now },
    endDate: { $gte: now },
  })
    .sort({ createdAt: -1 })
    .limit(limit);
};

ChallengeSchema.statics.findByCategory = function (category) {
  const now = new Date();
  return this.find({
    isActive: true,
    category,
    startDate: { $lte: now },
    endDate: { $gte: now },
  }).sort({ difficulty: 1, createdAt: -1 });
};

ChallengeSchema.statics.findByDifficulty = function (difficulty) {
  const now = new Date();
  return this.find({
    isActive: true,
    difficulty,
    startDate: { $lte: now },
    endDate: { $gte: now },
  }).sort({ featured: -1, createdAt: -1 });
};

ChallengeSchema.statics.getPopularChallenges = function (limit = 10) {
  return this.find({ isActive: true })
    .sort({ "stats.totalParticipants": -1, featured: -1 })
    .limit(limit);
};

ChallengeSchema.statics.getChallengeStats = function () {
  return this.aggregate([
    {
      $match: { isActive: true },
    },
    {
      $group: {
        _id: null,
        totalChallenges: { $sum: 1 },
        totalParticipants: { $sum: "$stats.totalParticipants" },
        averageCompletionRate: { $avg: "$stats.completionRate" },
        totalSavings: { $sum: { $toDouble: "$stats.averageSaving" } },
        challengesByCategory: {
          $push: {
            category: "$category",
            count: 1,
          },
        },
      },
    },
  ]);
};

const Challenge = mongoose.model("Challenge", ChallengeSchema);

export default Challenge;
